### PR TITLE
Add OpenSSL 3 support for dhchap-key generation

### LINKED ISSUE #1295: error: ‘HMAC_CTX_free’ is deprecated: Since OpenSSL 3.0
Looks like the recent OPENSSL changes were developed with a deprecated Open SSL library.

```
rhel-storage-08:nvme-cli(master) > make nvme
NVME_VERSION = 1.14.331.gf550
    CC ccan/tools/configurator/configurator
ccan/tools/configurator/configurator > ccan/config.h
    CC nvme.o
nvme.c: In function ‘gen_dhchap_key’:
nvme.c:6814:17: error: ‘HMAC_CTX_new’ is deprecated: Since OpenSSL 3.0 [-Werror=deprecated-declarations]
 6814 |                 HMAC_CTX *hmac_ctx = HMAC_CTX_new();
      |                 ^~~~~~~~
In file included from nvme.c:50:
/usr/include/openssl/hmac.h:33:33: note: declared here
   33 | OSSL_DEPRECATEDIN_3_0 HMAC_CTX *HMAC_CTX_new(void);
      |                                 ^~~~~~~~~~~~
nvme.c:6818:17: error: ‘ENGINE_load_builtin_engines’ is deprecated: Since OpenSSL 3.0 [-Werror=deprecated-declarations]
 6818 |                 ENGINE_load_builtin_engines();
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from nvme.c:48:
/usr/include/openssl/engine.h:358:28: note: declared here
  358 | OSSL_DEPRECATEDIN_3_0 void ENGINE_load_builtin_engines(void);
      |                            ^~~~~~~~~~~~~~~~~~~~~~~~~~~
nvme.c:6819:17: error: ‘ENGINE_register_all_complete’ is deprecated: Since OpenSSL 3.0 [-Werror=deprecated-declarations]
 6819 |                 ENGINE_register_all_complete();
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from nvme.c:48:
/usr/include/openssl/engine.h:415:27: note: declared here
  415 | OSSL_DEPRECATEDIN_3_0 int ENGINE_register_all_complete(void);
      |                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
nvme.c:6821:17: error: ‘HMAC_Init_ex’ is deprecated: Since OpenSSL 3.0 [-Werror=deprecated-declarations]
 6821 |                 HMAC_Init_ex(hmac_ctx, raw_secret, cfg.key_len,md, NULL);
      |                 ^~~~~~~~~~~~
In file included from nvme.c:50:
/usr/include/openssl/hmac.h:43:27: note: declared here
   43 | OSSL_DEPRECATEDIN_3_0 int HMAC_Init_ex(HMAC_CTX *ctx, const void *key, int len,
      |                           ^~~~~~~~~~~~
nvme.c:6822:17: error: ‘HMAC_Update’ is deprecated: Since OpenSSL 3.0 [-Werror=deprecated-declarations]
 6822 |                 HMAC_Update(hmac_ctx, (unsigned char *)hostnqn,
      |                 ^~~~~~~~~~~
In file included from nvme.c:50:
/usr/include/openssl/hmac.h:45:27: note: declared here
   45 | OSSL_DEPRECATEDIN_3_0 int HMAC_Update(HMAC_CTX *ctx, const unsigned char *data,
      |                           ^~~~~~~~~~~
nvme.c:6824:17: error: ‘HMAC_Update’ is deprecated: Since OpenSSL 3.0 [-Werror=deprecated-declarations]
 6824 |                 HMAC_Update(hmac_ctx, (unsigned char *)hmac_seed,
      |                 ^~~~~~~~~~~
In file included from nvme.c:50:
/usr/include/openssl/hmac.h:45:27: note: declared here
   45 | OSSL_DEPRECATEDIN_3_0 int HMAC_Update(HMAC_CTX *ctx, const unsigned char *data,
      |                           ^~~~~~~~~~~
nvme.c:6826:17: error: ‘HMAC_Final’ is deprecated: Since OpenSSL 3.0 [-Werror=deprecated-declarations]
 6826 |                 HMAC_Final(hmac_ctx, key, &key_len);
      |                 ^~~~~~~~~~
In file included from nvme.c:50:
/usr/include/openssl/hmac.h:47:27: note: declared here
   47 | OSSL_DEPRECATEDIN_3_0 int HMAC_Final(HMAC_CTX *ctx, unsigned char *md,
      |                           ^~~~~~~~~~
nvme.c:6827:17: error: ‘HMAC_CTX_free’ is deprecated: Since OpenSSL 3.0 [-Werror=deprecated-declarations]
 6827 |                 HMAC_CTX_free(hmac_ctx);
      |                 ^~~~~~~~~~~~~
In file included from nvme.c:50:
/usr/include/openssl/hmac.h:35:28: note: declared here
   35 | OSSL_DEPRECATEDIN_3_0 void HMAC_CTX_free(HMAC_CTX *ctx);
      |                            ^~~~~~~~~~~~~
cc1: all warnings being treated as errors
make: *** [Makefile:152: nvme.o] Error 1

rhel-storage-08:nvme-cli(master) > rpm -q --whatprovides /usr/include/openssl/hmac.h
openssl-devel-3.0.0-4.el9.x86_64

rhel-storage-08:nvme-cli(master) > git blame nvme.c | grep -C 2 HMAC_CTX_new
a05d4213 (Hannes Reinecke                  2020-12-17 09:43:50 +0100 6812)      if (md) {
a05d4213 (Hannes Reinecke                  2020-12-17 09:43:50 +0100 6813) #ifdef OPENSSL
a05d4213 (Hannes Reinecke                  2020-12-17 09:43:50 +0100 6814)              HMAC_CTX *hmac_ctx = HMAC_CTX_new();
a05d4213 (Hannes Reinecke                  2020-12-17 09:43:50 +0100 6815)              const char hmac_seed[] = "NVMe-over-Fabrics";
a05d4213 (Hannes Reinecke                  2020-12-17 09:43:50 +0100 6816)              unsigned int key_len;
```

### PR DESCRIPTION
**Pull request by igaw**
nvme: Add OpenSSL 3 support for DH-HMAC-CHAP key generation

OpenSSL 3 deprecated the HMAC_* API and introduced a new high level
API (EVP_*). Unfortunately, the EVP_* API is not available in OpenSSL
1.x. Add support for OpenSSL 3 as alternative implementation to
generate the DH-HMAC-CHAP key.

While at it also add error handling to the OpenSSL 1 version.

Signed-off-by: Daniel Wagner <dwagner@suse.de>

Fixes #1295

### CODE REVIEW

File: nvme.c [hunk 1]
```diff
@@ -6674,6 +6679,152 @@ static int show_hostnqn_cmd(int argc, char **argv, struct command *command, stru
 	return 0;
 }
 
+#ifdef CONFIG_OPENSSL
+#ifdef CONFIG_OPENSSL_1
+static int gen_hostname_hmac_ossl1(char *hostname, unsigned int hmac,
```
Comment (igaw) [inline nvme.c:6684]: I;ll update this s/hostname/hostnqn/

File: nvme.c [hunk 2]
```diff
@@ -6674,6 +6679,152 @@ static int show_hostnqn_cmd(int argc, char **argv, struct command *command, stru
...
+		md = EVP_sha384();
+		break;
+	case 3:
+		md = EVP_sha512();
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	HMAC_Init_ex(hmac_ctx, secret, key_len, md, NULL);
```
Comment (igaw) [inline nvme.c:6709]: And let me add some error checking for the 1.x version too

**Comment by johnmeneghini** [comment]
I've compiled your branch with both RHEL 8.5 and RHEL 9.0.  It all looks good to me.  Please merge this.

File: nvme.c [hunk 3]
```diff
@@ -6674,6 +6679,177 @@ static int show_hostnqn_cmd(int argc, char **argv, struct command *command, stru
...
+out:
+	EVP_MAC_CTX_free(mac_ctx);
+	EVP_MAC_free(mac);
+	OSSL_LIB_CTX_free(lib_ctx);
+
+	return err;
+}
+#endif /* !CONFIG_OPENSSL_3 */
+
 static int gen_dhchap_key(int argc, char **argv, struct command *command, struct plugin *plugin)
```
Comment (johnmeneghini) [inline nvme.c:6853]: Here too.  Doesn't this belong in libnvme?

File: nvme.c [hunk 4]
```diff
@@ -6674,6 +6679,177 @@ static int show_hostnqn_cmd(int argc, char **argv, struct command *command, stru
 	return 0;
 }
 
+#ifdef CONFIG_OPENSSL_1
+static int gen_dhchap_key_ossl(char *hostnqn, unsigned int hmac,
```
Comment (johnmeneghini) [inline nvme.c:6683]: Only one comment: shouldn't this function go into libnvme?

**Review by johnmeneghini — APPROVED** [review]
Changes all look and and they compile on RHEL.  But don't these functions really belong in libnvme?

File: nvme.c (hunk 3 above)
Comment (igaw) [reply nvme.c:6853]: Yep, I agree. I just factored this part out of the existing nvme-cli code base without thinking on moving it to the library.

File: nvme.c (hunk 4 above)
Comment (igaw) [reply nvme.c]: Done.

File: nvme.c (hunk 3 above)
Comment (igaw) [reply nvme.c:6672]: Done.

**Comment by johnmeneghini** [comment]
I'm seeing the following errors when I clone this branch and attempt `meson .build`.

```
[59/62] Compiling C object nvme.p/nvme.c.o
../nvme.c: In function ‘gen_dhchap_key’:
../nvme.c:6799:6: warning: implicit declaration of function ‘nvme_gen_dhchap_key’; did you mean ‘gen_dhchap_key’? [-Wimplicit-function-declaration]
  if (nvme_gen_dhchap_key(cfg.nqn, cfg.hmac, cfg.key_len,
      ^~~~~~~~~~~~~~~~~~~
      gen_dhchap_key
[62/62] Linking target nvme
FAILED: nvme 
```
Here's what my configuration looks like.

```
jmeneghi:nvme-cli(openssl-3-support) > git -C libnvme log -1
commit 0e1d56b5d1f56b816cecb719682f37578cf85a87 (HEAD -> nvme_gen_dhchap_key, igaw/nvme_gen_dhchap_key)
Author: Daniel Wagner <dwagner@suse.de>
Date:   Thu Dec 23 14:59:01 2021 +0100

    linux: Add support to generate DH-HMAC-CHAP keys
    
    Initially the DH-HMAC-CHAP key generation code was part of the
    nvme-cli repository and depended on OpenSSL 1. Move the key generation
    code into the library. While at it also add support for OpenSSL 3.
    
    Signed-off-by: Daniel Wagner <dwagner@suse.de>
jmeneghi:nvme-cli(openssl-3-support) > git branch -vv
  master            57d9370 [upstream/master] Merge pull request #1301 from igaw/build-fix-include-config
* openssl-3-support fed5811 [igaw/openssl-3-support] Update libnvme.wrap
jmeneghi:nvme-cli(openssl-3-support) > git status
On branch openssl-3-support
Your branch is up to date with 'igaw/openssl-3-support'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   libnvme (new commits)

Untracked files:
  (use "git add <file>..." to include in what will be committed)
        .build/

no changes added to commit (use "git add" and/or "git commit -a")
```

**Review by johnmeneghini — APPROVED** [review]
While the meson build fails, using `make nvme` works with no problems.

```
jmeneghi:nvme-cli(openssl-3-support) > make clean; make nvme; nm nvme | grep nvme_gen_dhchap_key
0000000000454e90 T nvme_gen_dhchap_key
```
I think you should merge this and work on the meson build problem separately.

**Pull request merged**